server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -207,5 +215,53 @@ public SimpleHealthIndicatorDetails getDetails() {
                 )
             );
         }
+
+        public HealthIndicatorImpact getImpact() {
+            final HealthIndicatorImpact impact;
+            if (primaries.available == false && replicas.available == false) {
```

Comment:
As I was going through the impact messages I realised this check needs to inspect the indices and find the intersection of unavailable primaries and replicas. 

Currently, if primaries have unassigned shards for `my-index-0001` and replicas have unassigned shards for `my-index-0002` I believe this impact will incorrectly be reported.

What do you think @masseyke ?

Reply:
Oh good catch, you're right. Adding a unit test for that.

Reply:
This led me to returning a list of impacts rather than a single one, which had a few cascading effects (like merging SimpleHealthIndicatorImpact into HealthIndicatorImpact since there is no longer a need for EMPTY, so there are no longer multiple implementations of the interface).